src/core/Block.ts

```diff
@@ -0,0 +1,260 @@
+  }
+
+  protected getStateFromProps(props: any): void {
+    this.state = {};
+  }
+
+  init() {
+    this._createResources();
+    this.eventBus().emit(Block.EVENTS.FLOW_RENDER, this.props);
+  }
+
+  _componentDidMount(props: P) {
+    this.componentDidMount(props);
+  }
+
+  componentDidMount(props: P) {}
+
+  _componentDidUpdate(oldProps: P, newProps: P) {
+    const response = this.componentDidUpdate(oldProps, newProps);
+    if (!response) {
+      return;
+    }
+    this._render();
+  }
+
+  componentDidUpdate(oldProps: P, newProps: P) {
+    return true;
+  }
+
+  setProps = (nextProps: P) => {
+    if (!nextProps) {
+      return;
+    }
+
+    Object.assign(this.props, nextProps);
+  };
+
+  setChildProps = (childRefName: string, nextProps: P) => {
+
+    if (!nextProps || !childRefName) {
+      return;
+    }
+    const childComponent = this.retrieveChildByRef(childRefName);
+    childComponent.setProps(nextProps);
+    this.refs[childRefName] = childComponent.getContent();
+  };
+
+  retrieveChildByRef = (ref: string) => {
+
+    const childBlocks = Object.values(this.children).filter(
+      (c) => c.element === this.refs[ref]
+    );
+    if (childBlocks.length !== 1) {
+      console.warn(
+        `1 Ref with Name ${ref} is expected but was: ${childBlocks}`
+      );
+    }
+    return childBlocks[0];
+  };
+
+  setState = (nextState: any) => {
+    if (!nextState) {
+      return;
+    }
+
+    Object.assign(this.state, nextState);
+  };
+
+  get element() {
+    return this._element;
+  }
+
+  _render() {
+    const fragment = this._compile();
+
+    this._removeEvents();
+    const newElement = fragment.firstElementChild!;
+
+    this._element!.replaceWith(newElement);
+
+    this._element = newElement as HTMLElement;
+    this._addEvents();
+  }
+
+  protected render(): string {
+    return '';
+  }
+
+  getContent(): HTMLElement {
+    // Хак, чтобы вызвать CDM только после добавления в DOM
+    if (this.element?.parentNode?.nodeType === Node.DOCUMENT_FRAGMENT_NODE) {
+      setTimeout(() => {
+        if (
+          this.element?.parentNode?.nodeType !== Node.DOCUMENT_FRAGMENT_NODE
+        ) {
+          this.eventBus().emit(Block.EVENTS.FLOW_CDM);
+        }
+      }, 100);
+    }
+
+    return this.element!;
+  }
+
+  _makePropsProxy(props: any): any {
+    // Можно и так передать this
+    // Такой способ больше не применяется с приходом ES6+
+    const self = this;
+
+    return new Proxy(props as unknown as object, {
+      get(target: Record<string, unknown>, prop: string) {
+        const value = target[prop];
+        return typeof value === 'function' ? value.bind(target) : value;
+      },
+      set(target: Record<string, unknown>, prop: string, value: unknown) {
+        target[prop] = value;
+
+        // Запускаем обновление компоненты
+        // Плохой cloneDeep, в след итерации нужно заставлять добавлять cloneDeep им самим
+        self.eventBus().emit(Block.EVENTS.FLOW_CDU, { ...target }, target);
```

Comment:
Можно использовать стрелочные функции, тогда не придется прибегать к замыканию
```suggestion
    // Можно и так передать this
    // Такой способ больше не применяется с приходом ES6+
    const self = this;

    return new Proxy(props as unknown as object, {
      get(target: Record<string, unknown>, prop: string) {
        const value = target[prop];
        return typeof value === 'function' ? value.bind(target) : value;
      },
      set: (target: Record<string, unknown>, prop: string, value: unknown) => {
        target[prop] = value;

        // Запускаем обновление компоненты
        // Плохой cloneDeep, в след итерации нужно заставлять добавлять cloneDeep им самим
        this.eventBus().emit(Block.EVENTS.FLOW_CDU, { ...target }, target);
```